src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,152 @@
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException("Incorrect index: " + index);
+        }
+        Node<T> newNode = new Node<>(value);
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            addElementToBeginning(newNode);
+        } else {
+            addElementAtIndex(newNode,index);
+        }
+    }
+
+    private void addElementToBeginning(Node<T> newNode) {
+        newNode.next = head;
+        head.prev = newNode;
+        head = newNode;
+        size++;
+    }
+
+    private void addElementAtIndex(Node<T> newNode,int index) {
+        Node<T> current = getNodeByIndex(index);
+        newNode.next = current;
+        newNode.prev = current.prev;
+        current.prev.next = newNode;
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeByIndex(index);
+        return (T) current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeByIndex(index);
+        T removedValue = current.value;
+        current.value = value;
+        return removedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedElement = unlink(getNodeByIndex(index));
+        size--;
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.value != null && current.value.equals(object) || current.value == object) {
+                unlink(current);
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Incorrect index: " + index);
+        }
+    }
+
+    private T unlink(Node<T> current) {
+        if (current == head) {
+            head = head.next;
+            return (T) current.value;
+        } else if (current == tail) {
+            tail = tail.prev;
+            return (T) current.value;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        return (T) current.value;
```

Comment:
See the comment above. You don't really need to implement return statement in every if-else blocks. Also add size--;